File: solution/backtracking/15654/main.py
# Authored by : tony9402
# Co-authored by : -
# Link : http://boj.kr/fa5e88e6aac14dbabbbb8b5bc7037963
import sys

def input():
    return sys.stdin.readline().rstrip()

N, M = map(int, input().split())
arr  = sorted(list(map(int, input().split())))
choose = [ 0 for _ in range(10) ]
used   = [ 0 for _ in range(10) ]

def dfs(cnt):
    global N, M
    if cnt == M:
        for idx in range(cnt):
            print(arr[choose[idx]], end=' ')
        print()
        return

    for i in range(0,N):
        if used[i]:
            continue
        used[i] = 1
        choose[cnt] = i
        dfs(cnt + 1)
        used[i] = 0

dfs(0)

File: solution/backtracking/15654/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/3955b249d38d4c79851d4cf97d474c0b
#include<bits/stdc++.h>

using namespace std;

int choose[10];
int arr[10], N, M;
bool used[10];

void dfs(int cnt) {
    if(cnt == M) {
        for(int i=0;i<M;i++) {
            cout << arr[choose[i]] << '\n';
        }
        return;
    }

    for(int i=0;i<N;i++){
        if(used[i]) continue;
        used[i] = true;
        choose[cnt] = i;
        dfs(cnt + 1);
        used[i] = false;
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> N >> M;
    for(int i=0;i<N;i++) cin >> arr[i];
    sort(arr, arr + N);
    dfs(0);

    return 0;
}
